### PR TITLE
8297600: Check current thread in selected JRT_LEAF methods

### PR DESCRIPTION
**Pull request by shipilev**
With [JDK-8275286](https://bugs.openjdk.org/browse/JDK-8275286), we added the `Thread::current()` checks for most of the JRT entries. But `JRT_LEAF` is still not checked, because not every `JRT_LEAF` carries a `JavaThread` argument. Having assertions there helps for two reasons. First, these methods can be called from the stub/compiler code, which might be erroneous with thread handling (especially in x86_32 that does not have a dedicated thread register). Second, in the post-Loom world, current thread can change suddenly, as evidenced here: https://mail.openjdk.org/pipermail/hotspot-compiler-dev/2022-November/060779.html.

We can add the thread checks to relevant `JRT_LEAF` methods that accept `JavaThread*` too.

Additional testing:
 - [x] Linux x86_64 fastdebug `tier1`
 - [x] Linux x86_64 fastdebug `tier2`
 - [x] Linux x86_32 fastdebug `tier1`
 - [x] Linux x86_32 fastdebug `tier2`

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8297600](https://bugs.openjdk.org/browse/JDK-8297600): Check current thread in selected JRT_LEAF methods

### Reviewers
 * [David Holmes](https://openjdk.org/census#dholmes) (@dholmes-ora - **Reviewer**)
 * [Coleen Phillimore](https://openjdk.org/census#coleenp) (@coleenp - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk pull/11359/head:pull/11359` \
`$ git checkout pull/11359`

Update a local copy of the PR: \
`$ git checkout pull/11359` \
`$ git pull https://git.openjdk.org/jdk pull/11359/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 11359`

View PR using the GUI difftool: \
`$ git pr show -t 11359`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/11359.diff">https://git.openjdk.org/jdk/pull/11359.diff</a>

</details>

### CODE REVIEW

File: src/hotspot/share/jvmci/jvmciRuntime.cpp [hunk 1]
```diff
@@ -578,6 +581,8 @@ JRT_ENTRY(jlong, JVMCIRuntime::invoke_static_method_one_arg(JavaThread* current,
 JRT_END
 
 JRT_LEAF(void, JVMCIRuntime::log_object(JavaThread* thread, oopDesc* obj, bool as_string, bool newline))
+  assert(thread == JavaThread::current(), "pre-condition");
```
Comment (dholmes-ora) [inline src/hotspot/share/jvmci/jvmciRuntime.cpp:584]: `thread` seems unused in this function and so it is not obvious it has to be the current thread.

File: src/hotspot/share/jvmci/jvmciRuntime.cpp [hunk 2]
```diff
@@ -603,16 +608,20 @@ JRT_END
 #if INCLUDE_G1GC
 
 void JVMCIRuntime::write_barrier_pre(JavaThread* thread, oopDesc* obj) {
+  assert(thread == JavaThread::current(), "pre-condition");
```
Comment (dholmes-ora) [inline src/hotspot/share/jvmci/jvmciRuntime.cpp:611]: Not obvious thread is expected/required to be current

File: src/hotspot/share/jvmci/jvmciRuntime.cpp [hunk 3]
```diff
@@ -603,16 +608,20 @@ JRT_END
...
   G1BarrierSetRuntime::write_ref_field_pre_entry(obj, thread);
 }
 
 void JVMCIRuntime::write_barrier_post(JavaThread* thread, volatile CardValue* card_addr) {
+  assert(thread == JavaThread::current(), "pre-condition");
```
Comment (dholmes-ora) [inline src/hotspot/share/jvmci/jvmciRuntime.cpp:616]: Not obvious thread is expected/required to be current

**Review by dholmes-ora — COMMENTED** [review]
Unclear for many JVMCI functions that the thread argument is actually intended/required to be the current thread. It seems unused in many cases so why is it passed?

**Comment by shipilev** [comment]
> Unclear for many JVMCI functions that the thread argument is actually intended/required to be the current thread. It seems unused in many cases so why is it passed?

Yes, I agree the initial patch over-reached in some places. Please see new commit, which reduces it. I left `thread == JavaThread::current()` checks where I can argue the threads are expected to be current.

**Review by dholmes-ora — APPROVED** [review]
That seems more recognisably correct. It would be even better if `thread` were named `current` in those cases where it must be, but that is a separate RFE.

Thanks.

**Review by coleenp — APPROVED** [review]
I thought this was going to be in the JRT_LEAF macro in interfaceSupport.inline.hpp but this seems fine.

**Comment by shipilev** [comment]
Thanks! I am integrating then.

/integrate

**Pull request closed without merging**